eventuate-tram-view-support-rebuild/src/main/java/io/eventuate/tram/viewsupport/rebuild/DomainSnapshotExportService.java

```diff
@@ -143,23 +167,36 @@ private RecordMetadata getRecordMetadataFromFuture(CompletableFuture<?> completa
               "",
               JSonMapper.toJson(message));
 
-      metadata.add(recordInfo);
+      metadata.put(anchorMessageId, recordInfo);
     }
 
     return metadata
+            .entrySet()
             .stream()
-            .map(this::getRecordMetadataFromFuture)
-            .map(recordMetadata ->
-                    new TopicPartitionOffset(recordMetadata.topic(), recordMetadata.partition(), recordMetadata.offset()))
+            .map(e -> {
+              RecordMetadata recordMetadata = getRecordMetadataFromFuture(e.getValue());
+              return new TopicPartitionOffsetMessageId(recordMetadata.topic(), recordMetadata.partition(), recordMetadata.offset(), e.getKey());
+            })
             .collect(Collectors.toList());
   }
 
-  private void publishDomainEntity(T domainEntity) {
+  private void publishDomainEntity(T domainEntity, Map<Integer, TopicPartitionOffsetMessageId> topicPartitionOffsetMessageIds) {
     DomainEventWithEntityId domainEventWithEntityId = domainEntityToDomainEventConverter.apply(domainEntity);
 
+    int partition = eventuateKafkaProducer.partitionFor(domainClass.getName(), domainEventWithEntityId.getEntityId().toString());
+
+    TopicPartitionOffsetMessageId topicPartitionOffsetMessageId = topicPartitionOffsetMessageIds.get(partition);
+
+    String messageId = idGenerator
```

Comment:
I think it would be valuable to define an ViewSupportIdGenerator class that implemented all of this logic.

This code would just call idGenerator.genId() and not worry about the details.


```

idGenerator
 .incrementIdIfPossible(Int128.fromString(topicPartitionOffsetMessageId.getDatabaseId()))
            .map(Int128::asString)
            .orElseGet(this::generateAnchorMessageId);

private String generateAnchorMessageId() {
    return eventuateCommonJdbcOperations.insertIntoMessageTable(idGenerator,
              "", "", Collections.emptyMap(), eventuateSchema, true);
```

Reply:
ok